security/tasks.py

```diff
@@ -57,29 +76,90 @@ def on_start(self, task_id, args, kwargs):
+            error_message=einfo,
+            output=self.request.output_stream.getvalue(),
+            retries=self.request.retries
+        )
+
+    def expire_task(self, task):
+        task.change_and_save(
+            state=CeleryTaskLogState.EXPIRED,
+            stop=now(),
+            error_message='Task execution was expired by command'
+        )
+
     def on_success(self, retval, task_id, args, kwargs):
         if retval:
             self.request.output_stream.write('Return value is "{}"'.format(retval))
 
         self.get_task().change_and_save(
             state=CeleryTaskLogState.SUCCEEDED,
             stop=now(),
-            output=self.request.output_stream.getvalue()
+            output=self.request.output_stream.getvalue(),
+            retries=self.request.retries
         )
         self._call_callback('success', task_id, args, kwargs)
 
     def on_failure(self, exc, task_id, args, kwargs, einfo):
         try:
-            self.get_task().change_and_save(
+            task = self.get_task()
+            LOGGER.log(self.logger_level, self.retry_error_message.format(
+                attempt=self.request.retries, exception=str(exc), task=task, task_name=task.name, **kwargs
+            ))
+            task.change_and_save(
                 state=CeleryTaskLogState.FAILED,
                 stop=now(),
                 error_message=einfo,
-                output=self.request.output_stream.getvalue()
+                output=self.request.output_stream.getvalue(),
+                retries=self.request.retries
             )
         except CeleryTaskLog.DoesNotExist:
             pass
         self._call_callback('failure', task_id, args, kwargs)
 
+    def _get_eta(self, options):
+        if options.get('countdown') is not None:
+            return now() + timedelta(seconds=options['countdown'])
+        elif options.get('eta'):
+            return options['eta']
+        else:
+            return now()
+
+    def _get_time_limit(self, options):
+        if options.get('time_limit') is not None:
+            return options['time_limit']
+        elif self.soft_time_limit is not None:
+            return self.soft_time_limit
+        else:
+            return settings.CELERYD_TASK_SOFT_TIME_LIMIT
+
+    def _get_stale_time_limit(self, options):
+        if options.get('stale_time_limit') is not None:
+            return options['stale_time_limit']
+        elif self.stale_time_limit is not None:
+            return self.stale_time_limit
+        else:
+            return settings.CELERYD_TASK_STALE_TIME_LIMIT
+
+    def _get_expires(self, options):
+        if options.get('expires') is not None:
+            return options['expires']
+        elif self.expires is not None:
+            return self.expires
+        elif hasattr(settings, 'CELERYD_TASK_STALE_TIME_LIMIT'):
+            return now() + timedelta(
+                seconds=(self._get_stale_time_limit(options) * 60) - self._get_time_limit(options)
```

Comment:
Chápu, proč tam odečítáš `self._get_time_limit(options)`, ale moc mi to nedává smysl, protože když bude soft_time_limit nastavený na hodinu a task se stihne vykonat za minutu, tak mu zbytečně nastavíme brzký deadline, do kdy se musí dokončit. Ale je to asi jedno v případě, že `_get_stale_time_limit` bude vracet dostatečně velké číslo :-)

Reply:
Ale by se tady ale mělo kontrolovat, že `_get_stale_time_limit` > `_get_time_limit`.

Reply:
ano nad tou kontrolou jsem také přemýšlel, ale pak jsem se kouknul do kódu celery a došel jsem ke stejnému závěru. Když jsi trouba a neumíš se zamyslet nad tím jak časy nastavuješ tak se ti task prostě nespustí :D